File: kotlin/allergies/src/main/kotlin/Allergies.kt
class Allergies(private var score: Int) {
    var allergies = mutableSetOf<Allergen?>()

    private val orderedAllergens = hashMapOf(
        Allergen.EGGS.score to Allergen.EGGS,
        Allergen.PEANUTS.score to Allergen.PEANUTS,
        Allergen.SHELLFISH.score to Allergen.SHELLFISH,
        Allergen.STRAWBERRIES.score to Allergen.STRAWBERRIES,
        Allergen.TOMATOES.score to Allergen.TOMATOES,
        Allergen.CHOCOLATE.score to Allergen.CHOCOLATE,
        Allergen.POLLEN.score to Allergen.POLLEN,
        Allergen.CATS.score to Allergen.CATS
    )

    init {
        val keys = orderedAllergens.keys.sorted().reversed()
        keys.forEach { allergenScore ->
            val allergenHitScore = score.div(allergenScore)
            if (allergenHitScore >= 1) {
                allergies.add(orderedAllergens[allergenScore])
                score = calculateRemainingScore(allergenScore, allergenHitScore)
            }
        }
    }

    private fun calculateRemainingScore(key: Int, allergenHitScore: Int): Int {
        return if (score.rem(key) == 0) {
            0
        } else {
            score - (allergenHitScore * orderedAllergens[key]!!.score)
        }
    }

    fun isAllergicTo(allergen: Allergen): Boolean {
        return allergies.contains(allergen)
    }

    fun getList(): List<Allergen> {
        return allergies.reversed().filterNotNull()
    }

}

File: csharp/allergies/Allergies.cs
using System;
using System.Collections.Generic;

public enum Allergen
{
    Eggs = 1,
    Peanuts = 2,
    Shellfish = 4,
    Strawberries = 8,
    Tomatoes = 16,
    Chocolate = 32,
    Pollen = 64,
    Cats = 128
}

public class Allergies
{
    public Allergies(int mask)
    {
        this.mask = (Allergen)mask;
    }

    public Allergen mask { get; }

    public bool IsAllergicTo(Allergen allergen)
    {
        return (mask & allergen) == allergen;
    }

    public Allergen[] List()
    {
        var results = new List<Allergen>();

        foreach (Allergen allergen in Enum.GetValues(typeof(Allergen)))
        {
            results = AddIfAllergicTo(results, allergen);
        }

        return results.ToArray();
    }

    private List<Allergen> AddIfAllergicTo(List<Allergen> results, Allergen allergen)
    {
        if (IsAllergicTo(allergen))
        {
            results.Add(allergen);
        }

        return results;
    }
}